app/services/user.js

```diff
@@ -0,0 +1,58 @@
+const userRepository = require('../repositories/user');
+const error = require('../services/error');
+
+class UserService {
+  async getUser(id) {
+    const foundUser = await userRepository.getUser(id);
+    if (!foundUser || foundUser.deletedAt) {
+      error.throwNotFoundError("User not found.");
+    }
+    return foundUser;
+  }
+
+  async createUser(user) {
+    let userExists;
+    try {
+      userExists = await userRepository.getUserByEmail(user.email);
+    } catch (err) {
+      error.throwValidationError("Email is required.");
+    }
+    if (!userExists) {
+      try {
+        return await userRepository.createUser(user);
+      } catch (err) {
+        error.throwValidationError("User data invalid.");
+      }
+    } else {
+      error.throwValidationError("Email already used.");
+    }
+  }
+
+  async deleteUser(id) {
+    await this.getUser(id);
+    await userRepository.deleteUser(id);
+    return await userRepository.getUserNonParanoid(id);
+  }
+
+  async updateUser(id, newUserData) {
+    await this.getUser(id);
+    if (newUserData.email) {
+      error.throwAccessNotGrantedError("You can not change the email.");
```

Comment:
This error shouldn't be used in these cases, but rather where there is a problem with the credentials of the user, when the access level granted by his credential is not enough to perform an operation. In this case it is a validation error. The email is not accepted for this action no matter what the access level of the user is.